src/registry.js

```diff
@@ -1,64 +1,187 @@
+      return;
     }
-    return fn(registry, outletId, ...args);
+
+    // Add the inlet to the inlet map and a new Conduit for every outlet with the same label
+    this.assocInlet(inlet);
+    this.wireInlet(inlet);
   };
 
-export const watchOutlet = ensureOutletInitialized((registry, outletId, fn) => {
-  const watcherId = uuidV4();
-  registry.outlets[outletId].watchers[watcherId] = fn;
+  unregisterInlet = inlet => {
+    // Do nothing if the Outlet is not registered
+    if (!this.findInlet(inlet.getId())) {
+      return;
+    }
 
-  return () => {
-    delete registry.outlets[outletId].watchers[watcherId];
+    this.dissocInlet(inlet);
+    this.unwireInlet(inlet);
   };
-});
 
-export const mergeInletChildren = ({ outlets, children }, outletId) =>
-  createFragment(
-    Array.from((outlets[outletId] && outlets[outletId].inlets) || [])
-      .sort((i1, i2) => i1.index - i2.index)
-      .map(inlet => [inlet.id, children[inlet.id]])
-      .reduce((acc, [id, children]) => ({ ...acc, [id]: children }), {}),
-  );
+  // full inlet rewire
+  // useful when labels change
+  rewireInlet = inlet => {
+    this.unregisterInlet(inlet);
+    this.registerInlet(inlet);
+  };
 
-export const addInlet = ensureOutletInitialized((registry, outletId, inlet) => {
-  registry.outlets[outletId].inlets.add(inlet);
-  notifyWatchers(registry, outletId);
-});
+  registerOutlet = outlet => {
+    // Do not add a new Outlet if it is already registered
+    if (this.findOutlet(outlet.getId())) {
+      return;
+    }
+    // Add the outlet to the outlet map and a new conduit for every inlet with the same label
+    this.assocOutlet(outlet);
+    this.wireOutlet(outlet);
+  };
 
-export const removeInlet = ensureOutletInitialized((registry, outletId, inletId) => {
-  registry.outlets[outletId].inlets.forEach(inlet => {
-    if (inlet.id === inletId) {
-      registry.outlets[outletId].inlets.delete(inlet);
+  unregisterOutlet = outlet => {
+    // Do nothing if the Outlet is not registered
+    if (!this.findOutlet(outlet.getId())) {
+      return;
     }
-  });
-  notifyWatchers(registry, outletId);
-});
 
-export const updateChildren = (registry, inletId, children) => {
-  registry.children[inletId] = children;
+    this.dissocOutlet(outlet);
+    this.unwireOutlet(outlet);
+  };
+
+  // full outlet rewire
+  // useful when labels change
+  rewireOutlet = outlet => {
+    this.unregisterOutlet(outlet);
+    this.registerOutlet(outlet);
+  };
+
+  // force a render for all outlets connected with a given inlet
+  updateConduits = inlet => this.findConduitsByInlet(inlet).forEach(conduit => conduit.update());
+
+  // build children using all inlets for the given label
+  mergeChildrenForLabel = label =>
+    createFragment(
+      this.findInletsByLabel(label)
+        .sort((i1, i2) => i1.index - i2.index)
+        .map(inlet => [inlet.getId(), inlet.getChildren()])
+        .reduce((acc, [id, children]) => ({ ...acc, [id]: children }), {}),
```

Comment:
Do you need to sort the inlets? I mean, you are building a new object, which is unordered by definition (IIRC).